collect_app/src/main/java/org/odk/collect/android/widgets/UrlWidget.java

```diff
@@ -17,78 +17,88 @@
 
 /**
  * Widget that allows user to open URLs from within the form
  *
  * @author Yaw Anokwa (yanokwa@gmail.com)
  */
 @SuppressLint("ViewConstructor")
-public class UrlWidget extends QuestionWidget implements ButtonWidget {
+public class UrlWidget extends QuestionWidget {
 
-    private Uri uri;
-    final Button openUrlButton;
-    final TextView stringAnswer;
     private final CustomTabHelper customTabHelper;
 
-    public UrlWidget(Context context, QuestionDetails questionDetails) {
+    protected Button openUrlButton;
+    protected TextView stringAnswer;
+
+    public UrlWidget(Context context, QuestionDetails questionDetails, CustomTabHelper customTabHelper) {
         super(context, questionDetails);
+        this.customTabHelper = customTabHelper;
+    }
 
-        openUrlButton = createSimpleButton(getContext(), getFormEntryPrompt().isReadOnly(), context.getString(R.string.open_url), getAnswerFontSize(), this);
+    @Override
+    protected View onCreateAnswerView(Context context, FormEntryPrompt prompt, int answerFontSize) {
+        ViewGroup answerView = (ViewGroup) LayoutInflater.from(context).inflate(R.layout.url_widget_answer, null);
 
-        stringAnswer = getCenteredAnswerTextView(getContext(), getAnswerFontSize());
+        openUrlButton = answerView.findViewById(R.id.url_button);
+        stringAnswer = answerView.findViewById(R.id.url_answer_text);
 
-        String s = questionDetails.getPrompt().getAnswerText();
-        if (s != null) {
-            stringAnswer.setText(s);
-            uri = Uri.parse(stringAnswer.getText().toString());
+        if (prompt.isReadOnly()) {
+            openUrlButton.setVisibility(GONE);
+        } else {
+            openUrlButton.setTextSize(TypedValue.COMPLEX_UNIT_DIP, answerFontSize);
+            openUrlButton.setOnClickListener(v -> onButtonClick());
         }
 
-        // finish complex layout
-        LinearLayout answerLayout = new LinearLayout(getContext());
-        answerLayout.setOrientation(LinearLayout.VERTICAL);
-        answerLayout.addView(openUrlButton);
-        answerLayout.addView(stringAnswer);
-        addAnswerView(answerLayout, WidgetViewUtils.getStandardMargin(context));
+        stringAnswer.setTextSize(TypedValue.COMPLEX_UNIT_DIP, answerFontSize);
+        String answerText = prompt.getAnswerText();
+        if (answerText != null) {
+            stringAnswer.setText(answerText);
+        }
 
-        customTabHelper = new CustomTabHelper();
+        return answerView;
     }
 
     private boolean isUrlEmpty(TextView stringAnswer) {
         return stringAnswer == null || stringAnswer.getText() == null
                 || stringAnswer.getText().toString().isEmpty();
     }
 
+    private Uri getUri() {
+        return Uri.parse(stringAnswer.getText().toString());
+    }
+
     @Override
     public void clearAnswer() {
         Toast.makeText(getContext(), "URL is readonly", Toast.LENGTH_SHORT).show();
```

Comment:
We have `ToastUtils` class which we use for displaying toast so we should use it here as well.